File: ThalesMobileSimulator/ThalesCoreSource/HostCommands/BuildIn/ImportKey_A6.vb
''
'' This program is free software; you can redistribute it and/or modify
'' it under the terms of the GNU General Public License as published by
'' the Free Software Foundation; either version 2 of the License, or
'' (at your option) any later version.
''
'' This program is distributed in the hope that it will be useful,
'' but WITHOUT ANY WARRANTY; without even the implied warranty of
'' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'' GNU General Public License for more details.
''
'' You should have received a copy of the GNU General Public License
'' along with this program; if not, write to the Free Software
'' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'' 

Imports ThalesSim.Core.Message
Imports ThalesSim.Core.Cryptography
Imports ThalesSim.Core

Namespace HostCommands.BuildIn

    ''' <summary>
    ''' Imports a key encrypted under a ZMK.
    ''' </summary>
    ''' <remarks>
    ''' This class implements the A6 Racal command.
    ''' </remarks>
    <ThalesCommandCode("A6", "A7", "", "Imports a key encrypted under a ZMK")> _
    Public Class ImportKey_A6
        Inherits AHostCommand

        Private _keyType As String = ""
        Private _zmk As String = ""
        Private _key As String = ""
        Private _lmkScheme As String = ""
        Private _targetKeyScheme As KeySchemeTable.KeyScheme

        ''' <summary>
        ''' Constructor.
        ''' </summary>
        ''' <remarks>
        ''' The constructor sets up the A6 message parsing fields.
        ''' </remarks>
        Public Sub New()
            ReadXMLDefinitions()
        End Sub

        ''' <summary>
        ''' Parses the request message.
        ''' </summary>
        ''' <remarks>
        ''' This method parses the command message. The message header and message command
        ''' code are <b>not</b> part of the message.
        ''' </remarks>
        Public Overrides Sub AcceptMessage(ByVal msg As Message.Message)
            XML.MessageParser.Parse(msg, XMLMessageFields, kvp, XMLParseResult)
            If XMLParseResult = ErrorCodes.ER_00_NO_ERROR Then
                _keyType = kvp.Item("Key Type")
                _zmk = kvp.ItemCombination("ZMK Scheme", "ZMK")
                _key = kvp.ItemCombination("Key Scheme", "Key")
                _lmkScheme = kvp.Item("Key Scheme LMK")
                If kvp.ItemOptional("Key Scheme") <> "" Then
                    _targetKeyScheme = KeySchemeTable.GetKeySchemeFromValue(kvp.ItemOptional("Key Scheme"))
                Else
                    _targetKeyScheme = KeySchemeTable.KeyScheme.Unspecified
                End If
            End If
        End Sub

        ''' <summary>
        ''' Creates the response message.
        ''' </summary>
        ''' <remarks>
        ''' This method creates the response message. The message header and message reply code
        ''' are <b>not</b> part of the message.
        ''' </remarks>
        Public Overrides Function ConstructResponse() As Message.MessageResponse
            Dim mr As New MessageResponse
            Dim _warnParity As Boolean = False

            Dim LMKKeyPair As LMKPairs.LMKPair, var As String = ""
            Dim lmk_ks As KeySchemeTable.KeyScheme

            If ValidateKeyTypeCode(_keyType, LMKKeyPair, var, mr) = False Then Return mr
            If ValidateKeySchemeCode(_lmkScheme, lmk_ks, mr) = False Then Return mr

            If ValidateFunctionRequirement(KeyTypeTable.KeyFunction.Export, LMKKeyPair, var, mr) = False Then Return mr

            Dim cryptZMK As New HexKey(_zmk)
            Dim clearZMK As String = Utility.DecryptUnderLMK(cryptZMK.ToString, cryptZMK.Scheme, LMKPairs.LMKPair.Pair04_05, var)
            If Utility.IsParityOK(clearZMK, Utility.ParityCheck.OddParity) = False Then
                mr.AddElement(ErrorCodes.ER_10_SOURCE_KEY_PARITY_ERROR)
                Return mr
            End If

            'Changed to use the DecryptUnderZMK method instead of directly performing 3DES decrypt.
            'See http://thalessim.codeplex.com/Thread/View.aspx?ThreadId=217215.
            Dim clearKey As String = DecryptUnderZMK(clearZMK, Utility.RemoveKeyType(_key), _targetKeyScheme)
            If Utility.IsParityOK(clearKey, Utility.ParityCheck.OddParity) = False Then
                _warnParity = True
            End If

            Dim cryptKey As String = Utility.EncryptUnderLMK(Utility.RemoveKeyType(clearKey), lmk_ks, LMKKeyPair, var)
            Dim checkValue As String = TripleDES.TripleDESEncrypt(New HexKey(clearKey), ZEROES)

            Log.Logger.MinorInfo("ZMK (clear): " + clearZMK)
            Log.Logger.MinorInfo("Key (clear): " + clearKey)
            Log.Logger.MinorInfo("Key (LMK): " + cryptKey)
            Log.Logger.MinorInfo("Check value: " + checkValue.Substring(0, 6))

            mr.AddElement(ErrorCodes.ER_00_NO_ERROR)
            mr.AddElement(cryptKey)
            mr.AddElement(checkValue.Substring(0, 6))

            Return mr

        End Function

    End Class

End Namespace

File: ThalesSim.Core/Commands/Host/Implementations/ImportKey_A6.cs
/*
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

using ThalesSim.Core.Cryptography;
using ThalesSim.Core.Cryptography.LMK;
using ThalesSim.Core.Message;
using ThalesSim.Core.Resources;
using ThalesSim.Core.Utility;

namespace ThalesSim.Core.Commands.Host.Implementations
{
    [ThalesHostCommand("A6", "A7", "Imports a key encrypted under a ZMK")]
    public class ImportKey_A6 : AHostCommand
    {
        private string _keyType;
        private string _zmk;
        private string _key;
        private string _lmkScheme;
        private string _atallaVariant;

        public ImportKey_A6()
        {
            ReadXmlDefinitions();
        }

        public override void AcceptMessage(StreamMessage message)
        {
            base.AcceptMessage(message);

            if (XmlParseResult != ErrorCodes.ER_00_NO_ERROR)
            {
                return;
            }

            _keyType = KeyValues.Item("Key Type");
            _zmk = KeyValues.ItemCombination("ZMK Scheme", "ZMK");
            _key = KeyValues.ItemCombination("Key Scheme", "Key");
            _lmkScheme = KeyValues.Item("Key Scheme LMK");
            _atallaVariant = KeyValues.ItemOptional("Atalla Variant");
        }

        public override StreamResponse ConstructResponse()
        {
            var mr = new StreamResponse();

            KeyTypeCode ktc = null;
            if (!ValidateKeyTypeCode(_keyType, mr, ref ktc))
            {
                return mr;
            }

            var lmk = KeyScheme.Unspecified;
            if (!ValidateKeySchemeCode(_lmkScheme, mr, ref lmk))
            {
                return mr;
            }

            if (!ValidateAuthStateRequirement(KeyFunction.Import, ktc.Pair, ktc.Variant, mr))
            {
                return mr;
            }

            var zmk = new HexKeyThales(new KeyTypeCode(0, LmkPair.Pair04_05), false, _zmk);
            if (!zmk.ClearKey.IsParityOk(Parity.Odd))
            {
                mr.Append(ErrorCodes.ER_10_SOURCE_KEY_PARITY_ERROR);
                return mr;
            }

            var clearKey = zmk.DecryptWithScheme(_key, _atallaVariant);
            var warnParity = !clearKey.IsParityOk(Parity.Odd);

            var newKey = new HexKeyThales(ktc, true, clearKey);
            var checkValue = newKey.CheckValue;

            Log.InfoFormat("ZMK (clear): {0}", zmk.ClearKey);
            Log.InfoFormat("Key (clear): {0}", clearKey);
            Log.InfoFormat("Key (LMK): {0}", newKey.Key);
            Log.InfoFormat("Check value: {0}", checkValue);
            if (warnParity)
            {
                Log.WarnFormat("Key {0} does not have odd parity (would be {1})", clearKey, clearKey.MakeParity(Parity.Odd));
            }

            mr.Append(ErrorCodes.ER_00_NO_ERROR);
            mr.Append(lmk == KeyScheme.DoubleLengthKeyVariant || lmk == KeyScheme.TripleLengthKeyVariant
                          ? newKey.KeyVariant
                          : newKey.KeyAnsi);
            mr.Append(newKey.CheckValue.Substring(0, 6));

            return mr;
        }
    }
}
